Sources/PhishingDetection/PhishingDetector.swift

```diff
@@ -0,0 +1,106 @@
+    }
+
+    public var errorUserInfo: [String: Any] {
+        switch self {
+        case .detected:
+            return [NSLocalizedDescriptionKey: "Phishing detected"]
+        }
+    }
+
+    public var rawValue: Int {
+        return self.errorCode
+    }
+}
+
+public protocol PhishingDetecting {
+	func isMalicious(url: URL) async -> Bool
+}
+
+public class PhishingDetector: PhishingDetecting {
+    let hashPrefixStoreLength: Int = 8
+    let hashPrefixParamLength: Int = 4
+	var apiClient: PhishingDetectionClientProtocol
+	var dataStore: PhishingDetectionDataSaving
+
+	public init(apiClient: PhishingDetectionClientProtocol, dataProvider: PhishingDetectionDataProviding, dataStore: PhishingDetectionDataSaving) {
+		self.apiClient = apiClient
+		self.dataStore = dataStore
+	}
+
+	func getMatches(hashPrefix: String) async -> Set<Match> {
+		return Set(await apiClient.getMatches(hashPrefix: hashPrefix))
+	}
+
+	func inFilterSet(hash: String) -> Set<Filter> {
+		return Set(dataStore.filterSet.filter { $0.hashValue == hash })
+	}
+
+	func matchesUrl(hash: String, regexPattern: String, url: URL, hostnameHash: String) -> Bool {
+		if hash == hostnameHash,
+		   let regex = try? NSRegularExpression(pattern: regexPattern, options: []) {
+			let urlString = url.absoluteString
+			let range = NSRange(location: 0, length: urlString.utf16.count)
+			return regex.firstMatch(in: urlString, options: [], range: range) != nil
+		}
+		return false
+	}
+
+	public func isMalicious(url: URL) async -> Bool {
+		guard let canonicalHost = url.canonicalHost() else { return false }
+        guard let canonicalUrl = url.canonicalURL() else { return false }
+		let hostnameHash = SHA256.hash(data: Data(canonicalHost.utf8)).map { String(format: "%02hhx", $0) }.joined()
+		let hashPrefix = String(hostnameHash.prefix(hashPrefixStoreLength))
+		if dataStore.hashPrefixes.contains(hashPrefix) {
+			// Check local filterSet first
+			let filterHit = inFilterSet(hash: hostnameHash)
+            for filter in filterHit where matchesUrl(hash: filter.hashValue, regexPattern: filter.regex, url: canonicalUrl, hostnameHash: hostnameHash) {
+                // Commented out to avoid PixelKit dependency: https://github.com/duckduckgo/BrowserServicesKit/pull/934
+				// PixelKit.fire(PhishingDetectionPixels.errorPageShown(clientSideHit: true))
+                return true
+            }
+			// If nothing found, hit the API to get matches
+            let hashPrefixParam = String(hostnameHash.prefix(hashPrefixParamLength))
+			let matches = await apiClient.getMatches(hashPrefix: hashPrefixParam)
+			for match in matches where matchesUrl(hash: match.hash, regexPattern: match.regex, url: canonicalUrl, hostnameHash: hostnameHash) {
+                // Commented out to avoid PixelKit dependency: https://github.com/duckduckgo/BrowserServicesKit/pull/934
+				// PixelKit.fire(PhishingDetectionPixels.errorPageShown(clientSideHit: false))
+				return true
+			}
+		}
+		return false
+	}
+}
```

Comment:
Maybe we can break it down a bit for example we could have a matchesUrl function used twice for the loops and a function for computing hash etc